docs/keps/20200619-federated-resource-status.md

```diff
@@ -0,0 +1,408 @@
+Users lack of proper visibility over the status of the federated
+resources. For instance, if you federated a deployment the federated status should
+report if the deployment failed or error at any time.
+
+A federated resource only reflects the status of propagation actions, but it doesn't
+reflect the status if whether the resource is running or failed.
+
+## Motivation
+
+Nowadays users have to connect to the kubefed clusters to be aware if the federated
+resources are healthy or not across clusters.
+
+### Goals
+
+* Quickly identify unhealthy federated resources by relying on the status of the federated resources.
+* Improve the troubleshooting of failures when propagating resources.
+
+
+## Proposal
+
+Kubefed reports a limited set of states for the federation of resources.
+
+
+```go
+CreationTimedOut     PropagationStatus = "CreationTimedOut"
+UpdateTimedOut       PropagationStatus = "UpdateTimedOut"
+DeletionTimedOut     PropagationStatus = "DeletionTimedOut"
+LabelRemovalTimedOut PropagationStatus = "LabelRemovalTimedOut"
+
+AggregateSuccess       AggregateReason = ""
+ComputePlacementFailed AggregateReason = "ComputePlacementFailed"
+NamespaceNotFederated  AggregateReason = "NamespaceNotFederated"
+
+PropagationConditionType ConditionType = "Propagation"
+```
+
+However the current federated resource properties don't help to track the status of the deployed resources in the kubefed clusters.
+
+The idea is to extend current GenericFederatedStatus with the Status of the resources:
+
+```go
+
+type GenericFederatedStatus struct {
+  	ObservedGeneration  int64                  `json:"observedGeneration,omitempty"`
+  	Conditions          []*GenericCondition    `json:"conditions,omitempty"`
+  	Clusters            []GenericClusterStatus `json:"clusters,omitempty"`
+}
+
+type GenericFederatedResource struct {
+	metav1.TypeMeta                `json:",inline"`
+	metav1.ObjectMeta              `json:"metadata,omitempty"`
+
+	Status *GenericFederatedStatus `json:"status,omitempty"`
+}
+
+```
+
+Nowadays `Conditions` hold the status of the federated actions (aka propagation status).
+In other words, it defines the conditions of the propagation status for a resource.
+
+```yaml
+- apiVersion: types.kubefed.io/v1beta1
+  kind: FederatedDeployment
+  metadata:
+    finalizers:
+    - kubefed.io/sync-controller
+    generation: 1
+    name: asystem
+    namespace: asystem
+    resourceVersion: "70174497"
+  spec:
+    placement:
+      clusters:
+      - name: cluster3
+      - name: cluster2
+      - name: cluster1
+    template:
+      metadata:
+        labels:
+          app: nginx
+      spec:
+        replicas: 3
+        selector:
+          matchLabels:
+            app: nginx
+        template:
+          metadata:
+            labels:
+              app: nginx
+          spec:
+            containers:
+            - image: nginx
+              name: nginx
+  status:
```

Comment:
```suggestion
  status:
    clusters:
      - name: cluster1
      - name: cluster2
      - name: cluster3
```

Reply:
I guess it's better to show all fields of `status` letting people know the current situations and get what will be changed.
Same as below.

Reply:
This is an example of a current status as you can see it today. I show the clusters and conditions field, what else you would like to see.